component-monitor/src/main.rs

```diff
@@ -0,0 +1,242 @@
+    info!(target: "files", "populating {target:?} from {source:?} with sentinel {sentinel_name:?}");
+    let sentinel_src = source.join(sentinel_name);
+
+    if !sentinel_src.exists() {
+        info!(target: "files", "sentinel file ({:?}) not found, skipping", &sentinel_src);
+        return Ok(());
+    }
+
+    let sentinel_data = match fs::read(&sentinel_src) {
+        Ok(d) if d.trim_ascii() == b"" => {
+            error!(target: "files", "found empty sentinel file ({:?})", &sentinel_src);
+            return Ok(());
+        }
+        Ok(d) => d,
+        Err(e) => {
+            error!(target: "files", "failed to read sentinel file: {e}");
+            return Ok(());
+        }
+    };
+
+    let sentinel_tgt = target.join(sentinel_name);
+    if fs::read(&sentinel_tgt).is_ok_and(|content| content == sentinel_data) {
+        info!(target: "files", "found current sentinel, skipping");
+        return Ok(());
+    }
+
+    cleanup(target, sentinel_name)?;
+    info!(target: "files", "copying from {source:?} to {target:?}");
+    for entry in WalkDir::new(source) {
+        let entry = match entry {
+            Ok(e) => e,
+            Err(e) => {
+                error!(target: "files", "{e}");
+                continue;
+            }
+        };
+
+        let path = entry.path();
+        let relative_path = path.strip_prefix(source)?;
+        if path == source {
+            continue;
+        }
+        if relative_path.as_os_str() == sentinel_name {
+            continue;
+        }
+
+        let target_path = target.join(relative_path);
+        if path.is_dir() {
+            if let Err(err) = fs::create_dir(&target_path) {
+                error!(target: "files", "failed to create {:?}: {err}", &target_path);
+            } else {
+                debug!(target: "files", "created {:?}", &target_path)
+            }
+        } else {
+            if let Err(err) = fs::copy(path, &target_path) {
+                error!(target: "files", "failed to copy {:?} to {:?}: {err}", &path, &target_path);
+            } else {
+                debug!(target: "files", "copied {:?} to {:?}", &path, &target_path);
+            }
+        }
+    }
+
+    if let Err(err) = fs::copy(&sentinel_src, &sentinel_tgt) {
+        error!(target: "files", "failed to copy sentinel: {err}");
+    } else {
+        debug!(target: "files", "copied sentinel ({:?} to {:?})", sentinel_src, sentinel_tgt);
+    }
+
+    Ok(())
+}
+
+/// Handle inotify events:
+/// - run cleanup on sentinel deletion
+/// - populate on sentinel written
+/// - panic on source remove
+fn monitor(
+    inotify: &Inotify,
+    source: &PathBuf,
+    sentinel_name: &str,
+    target: &PathBuf,
+) -> Result<()> {
+    info!(target: "inotify", "starting event monitoring on {:?}", &source);
+    loop {
+        let events = inotify
+            .read_events()
+            .context("Error reading Inotify events")?;
```

Comment:
The loop blocks on that `read_events`. If there's no activity, it might be better to implement a timeout or a way to check for shutdown signals.

Reply:
That's the purpose of this service, though… wait for activity in there… I'm not sure if the added complexity adds any value here.

Reply:
I guess the only benefit of having a shutdown signal here would be that, as it stands, stopping this service at the wrong time will leave an operation half-done. The delete-sentinel-first, write-sentinel-last approach will ensure that this *recovers*, but I don't have the context to know if leaving this temporarily in a broken state is acceptable. (I guess if you *really* wanted to avoid a temporary broken state you'd need to do this a bit differently anyway)